### ISSUE TITLE
[unitaryhack] QNodes no longer display readable info when printed

### ISSUE THREAD

**Issue by albi3ro** [PennyLaneAI/pennylane#1352]
*This issue has been tagged for contributions during [unitaryHACK](https://unitaryfund.github.io/unitaryhack/participating-projects.html)*

The [QNode](https://github.com/PennyLaneAI/pennylane/blob/0965bb0c97c4de2dc2408751ca3a61dd47074264/pennylane/qnode.py#L34) currently does not display useful information when printed out:

```
>>> qnode
<pennylane.qnode.QNode at 0x7f4e29173d00>
```

We should add `__str__`  and `__repr__` methods to the class to facilitate usage and debugging.

The `__str__` method needs to be easily human-readable, while the `__repr__`  method should be more comprehensive.  

Both methods should at least convey the qnode's:

* `diff_method`
* `interface`
* `device.short_name`
* `device.num_wires`

**Comment by josh146** [comment]
Thanks @albi3ro! For both the `__repr__` and the `__str__`, I would be in favour of something simple like

```python
<QNode: wires=[0, “a”, 2], device=“default.qubit”, interface=“tf”, diff_method=“adjoint”>
```

This matches how it used to display, and I would almost consider it a bug that this changed.

Note: if we don’t define `__str__`, Python will automatically default to `__repr__` when `print(qnode)` is called.

**Comment by paniash** [comment]
Hi! I'd like to try my hand at this. Any pointers for this? (pretty new to OOP in python). Thanks. :-)

**Comment by josh146** [comment]
Hi @paniash! This should be as simple as defining the QNode’s `__repr__` method, and having it return the string to be displayed.

In fact, you could have a look at the previous implementation: 

https://github.com/PennyLaneAI/pennylane/blob/4fffb1bb08bee0569c259e05f5ba59c0d2db103c/pennylane/qnodes/base.py#L207

**Comment by paniash** [comment]
@josh146 Thanks! I've already made a PR. Could you check it and let me know if it's alright? (I've only written the `__str__` method though)

**Comment by paniash** [comment]
Do we need to have both `__str__` and `__repr__` methods? I feel like both try to achieve the same thing.

**Comment by josh146** [comment]
Hey @paniash, there is a subtle difference between the two.

For example, consider

```python
class TestClass:
    def __init__(self, info):
        self.info = info
    def __str__(self):
        return f"Some detailed information: {self.info}"
    def __repr__(self):
        return f"<TestClass: {self.info}>"
```

* `__repr__` is used in interactive or console environments, such as the Python console, IPython, or Jupyter notebooks. This is what is displayed if you press enter after a command or a variable:

  ```pycon
  >>> x = TestClass(5)
  >>> x
  <TestClass: 5>
  ```

  It is a concise, unambiguous representation of the object, that can usually be used to reconstruct the object or used for debugging. If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value---however, this won't be possible with a QNode!

* `__str__` is used if `format()` or `print()` is called on the object:

  ```pycon
  >>> print(x)
  Some detailed information: 5
  ```

  It provides detailed, informal, human readable information for the end-user.

Note that if `__str__` is not defined, `__repr__` will be used when `print()` or `format()` is called.

**Comment by paniash** [comment]
@josh146 Gotcha. Thanks for the explanation! I'll make the changes accordingly in the PR. :-)

**Comment by paniash** [comment]
Fixed in [`be3c88d`](https://github.com/PennyLaneAI/pennylane/commit/be3c88d320ee48a0f3fb761c981356137de20027).
